core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowRewriter.scala

```diff
@@ -0,0 +1,424 @@
+      }
+      logger.info("Operator {} cache invalidated.", id)
+    }
+
+    def invalidateOperatorCacheRecursively(opID: String): Unit = {
+      invalidateOperatorCache(opID)
+      workflowDAG
+        .getDownstream(opID)
+        .foreach(desc => {
+          invalidateOperatorCacheRecursively(desc.operatorID)
+        })
+    }
+
+    def isUpdated(opID: String): Boolean = {
+      if (!operatorRecord.contains(opID)) {
+        operatorRecord += ((opID, getWorkflowVertex(workflowDAG.getOperator(opID))))
+        logger.info("Vertex {} is not recorded.", operatorRecord(opID))
+        true
+      } else if (workflowInfo.cachedOperatorIDs.contains(opID)) {
+        !operatorRecord(opID).equals(getWorkflowVertex(workflowDAG.getOperator(opID)))
+      } else {
+        val vertex = getWorkflowVertex(workflowDAG.getOperator(opID))
+        if (!operatorRecord(opID).equals(vertex)) {
+          operatorRecord(opID) = vertex
+          logger.info("Vertex {} is updated.", operatorRecord(opID))
+          true
+        } else if (cachedOperatorDescriptors.contains(opID)) {
+          !workflowInfo.cachedOperatorIDs.contains(opID)
+        } else {
+          logger.info("Operator: {} is not updated.", operatorRecord(opID))
+          false
+        }
+      }
+    }
+
+    def cacheStatusUpdateHelper(opID: String): Unit = {
+      if (visitedOpID.contains(opID)) {
+        return
+      }
+      visitedOpID += opID
+      logger.info(
+        "Checking update status of operator {}.",
+        workflowDAG.getOperator(opID).toString
+      )
+      if (isUpdated(opID)) {
+        invalidateOperatorCache(opID)
+        workflowDAG
+          .getDownstream(opID)
+          .map(op => op.operatorID)
+          .foreach(invalidateOperatorCacheRecursively)
+      }
+    }
+
+    val opIter = workflowDAG.jgraphtDag.iterator()
+    while (opIter.hasNext) {
+      val opID = opIter.next()
+      if (!visitedOpID.contains(opID)) {
+        cacheStatusUpdateHelper(opID)
+      }
+    }
+    workflowInfo.operators
+      .map(op => op.operatorID)
+      .filterNot(visitedOpID.contains)
+      .foreach(invalidSet.+=)
+    invalidSet
+  }
+
+  private def checkCacheValidity(): Unit = {
+    val opIter = workflowDAG.jgraphtDag.iterator()
+    while (opIter.hasNext) {
+      val id = opIter.next()
+      if (!visitedOpID.contains(id)) {
+        invalidateIfUpdated(id)
+      }
+    }
+  }
+
+  private def invalidateIfUpdated(operatorID: String): Unit = {
+    if (visitedOpID.contains(operatorID)) {
+      return
+    }
+    visitedOpID += operatorID
+    logger.info(
+      "Checking update status of operator {}.",
+      workflowDAG.getOperator(operatorID).toString
+    )
+    if (isUpdated(operatorID)) {
+      invalidateOperatorCache(operatorID)
+      workflowDAG
+        .getDownstream(operatorID)
+        .map(op => op.operatorID)
+        .foreach(invalidateOperatorCacheRecursively)
+    }
+  }
+
+  def isUpdated(id: String): Boolean = {
```

Comment:
ok now, this is another `isUpdated` method. I am confused. What is this one and what is the difference with the previous one?

Reply:
edu.uci.ics.texera.workflow.common.workflow.WorkflowRewriter#isUpdated in line 289 is for rewriter, and isUpdated in line 207 inside edu.uci.ics.texera.workflow.common.workflow.WorkflowRewriter#cacheStatusUpdate is used for cache status update. They do very similar thing with slight differences.